### PR TITLE
Align User Lookup implementation 

### LINKED ISSUE #5008: retrieveUserInfo/retrieveUserInfoAsync
User lookup was implemented in https://github.com/realm/realm-java/issues/4828 and https://github.com/realm/realm-java/pull/4882

Unfortunately, some miscommunication happened and 4 bindings actually managed to implement it in 4 different ways :man_facepalming: 

Long story short. We need to deprecate the current `retrieveUser` implementation and implement `retrieveUserInfo()` and `retrieveUserInfoAsync()` instead as described in: https://github.com/realm/product/issues/364#issuecomment-317110948

### PR DESCRIPTION
**Pull request by nhachicha**
fixes #5008

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 1]
```diff
@@ -444,7 +444,9 @@ public SyncUser run() {
      * @return {@code SyncUser} associated with the given identity provider and providerId, or {@code null} in case
      * of an {@code invalid} provider or {@code providerId}.
      * @throws ObjectServerError in case of an error.
+     * @deprecated  As of release 3.6.0, replaced by {@link #retrieveInfoForUser(String, String)}}
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: `As` -> `as`  lowercase for others, the first character following javadoc items.

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 2]
```diff
@@ -511,6 +515,80 @@ public SyncUser run() {
...
+     *                             In most cases this is different from the Realm Object Server-issued identity.
+     * @param provider The authentication provider {@link io.realm.SyncCredentials.IdentityProvider} that manages the user whose information is desired.
+     *
+     * @return {@code SyncUser} associated with the given identity provider and providerId, or {@code null} in case
+     * of an {@code invalid} provider or {@code providerId}.
+     * @throws ObjectServerError in case of an error.
+     */
+    public SyncUserInfo retrieveInfoForUser(final String providerUserIdentity, final String provider) throws ObjectServerError {
+        if (Util.isEmptyString(providerUserIdentity)) {
+            throw new IllegalArgumentException("Not-null 'providerUserIdentity' required.");
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: Not-null -> non empty?

Comment (Zhuinden) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: I tend to do `'providerUserIdentity' cannot be empty!` but it depends on convention.

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java (hunk 1 above)
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: Fine if we stick to this convention, it's upper case on the annotation [doc](https://docs.oracle.com/javase/8/docs/technotes/guides/javadoc/deprecation/deprecation.html) btw

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 3]
```diff
@@ -444,7 +444,7 @@ public SyncUser run() {
      * @return {@code SyncUser} associated with the given identity provider and providerId, or {@code null} in case
      * of an {@code invalid} provider or {@code providerId}.
      * @throws ObjectServerError in case of an error.
-     * @deprecated  As of release 3.6.0, replaced by {@link #retrieveInfoForUser(String, String)}}
+     * @deprecated  as of release 3.6.0, replaced by {@link #retrieveInfoForUser(String, String)}}
```
Comment (dalinaum) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: It seems two spaces between `@deprecated` and `as`. Can we use one space instead?

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 4]
```diff
@@ -497,7 +497,7 @@ public SyncUser retrieveUser(final String provider, final String providerId) thr
      * @param callback callback when the lookup has completed or failed. The callback will always happen on the same thread
      * as this method is called on.
      * @return representation of the async task that can be used to cancel it if needed.
-     * @deprecated  As of release 3.6.0, replaced by {@link #retrieveInfoForUserAsync(String, String, RequestCallback)}}
+     * @deprecated  as of release 3.6.0, replaced by {@link #retrieveInfoForUserAsync(String, String, RequestCallback)}}
```
Comment (dalinaum) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: Same above.

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 5]
```diff
@@ -444,7 +444,9 @@ public SyncUser run() {
      * @return {@code SyncUser} associated with the given identity provider and providerId, or {@code null} in case
      * of an {@code invalid} provider or {@code providerId}.
      * @throws ObjectServerError in case of an error.
+     * @deprecated  as of release 3.6.0, replaced by {@link #retrieveInfoForUser(String, String)}}
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: whitespace

File: realm/realm-library/src/objectServer/java/io/realm/SyncUserInfo.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+import io.realm.internal.network.LookupUserIdResponse;
+
+/**
+ * POJO representing information about a user that was retrieved from a user lookup call.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUserInfo.java]: Add a `@see <releavant method>`

File: realm/realm-library/src/objectServer/java/io/realm/SyncUserInfo.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+public class SyncUserInfo {//TODO use getter for Javadoc
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUserInfo.java]: Should we implement `equals/hashcode/toString` ?

**Pull request merged**